### PR TITLE
pyopae implementation

### PR DESCRIPTION
**Pull request by r-rojo**
This a re-implementation of #413 
This breaks up the C++ backing code into separate files to make it easier to read and also includes the Python docstrings.
Things to note in this implementation that are different from the original version are:
* `handle.reconfigure` uses a file object as one of the arguments. This calls the Python API to get a FILE* object and read the buffer. This allows opening/closing the GBS files from Python instead of delegating that to C++.
* `shared_buffer` implements the buffer protocol. This allows it to be used with `numpy` arrays.

### CODE REVIEW

File: pyopae/opae.cpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+      .def_property("socket_id", properties_get_socket_id,
+                    properties_set_socket_id, properties_doc_socket_id())
+      .def_property("object_id", properties_get_object_id,
+                    properties_set_object_id, properties_doc_object_id())
+      .def_property("num_slots", properties_get_num_slots,
+                    properties_set_num_slots, properties_doc_num_slots())
+      .def_property("bbs_id", properties_get_bbs_id, properties_set_bbs_id,
+                    properties_doc_bbs_id())
+      .def_property("bbs_version", properties_get_bbs_version,
+                    properties_set_bbs_id, properties_doc_bbs_id())
```
Comment (ibarraza) [inline pyopae/opae.cpp]: Shouldn't this be "properties_set_bbs_version, properties_doc_bbs_version())"?

File: pyopae/opae.cpp [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        py::arg("flags") = 0);
+  py::class_<handle, handle::ptr_t> pyhandle(m, "handle");
+  pyhandle
+      .def("__enter__", handle_context_enter, handle_doc_context_enter())
+      .def("__exit__", handle_context_exit, handle_doc_context_exit())
+      .def("reconfigure", handle_reconfigure, handle_doc_reconfigure(),
+           py::arg("slot"), py::arg("fd"), py::arg("flags") = 0)
+      .def("__bool__", handle_valid, handle_doc_valid())
+      .def("close", &handle::close, handle_doc_close())
+      .def("reset", &handle::reset)
```
Comment (ibarraza) [inline pyopae/opae.cpp]: Shouldn't this include handle_doc_reset()?

File: pyopae/pyhandle.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+             checking if the resource is currently in use.
+
+  )opaedoc";
+}
+
+void handle_reconfigure(handle::ptr_t handle, uint32_t slot, py::object file,
+                        int flags) {
+  PyObject *obj = file.ptr();
+  if (!PyFile_Check(obj)) {
+    throw std::invalid_argument("fd argument is not a file objeict");
```
Comment (ibarraza) [inline pyopae/pyhandle.cpp]: Typo.

File: pyopae/opae.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      "Flags that define how an accelerator is opened.")
+      .value("RECONF_FORCE", FPGA_RECONF_FORCE)
+      .export_values();
+
+  // define properties class
+  py::class_<properties, properties::ptr_t> pyproperties(m, "properties",
+                                                         properties_doc());
+  pyproperties.def(py::init(&properties_get), properties_doc_get())
+      .def(py::init(&properties_get_token), properties_doc_get_token())
+      .def_property("parent", properties_get_parent, properties_get_parent,
```
Comment (ibarraza) [inline pyopae/opae.cpp]: Shouldn't one of these be properties_set_parent?

**Comment by r-rojo** [comment]
@ibarraza Good 🐱ch

Thanks for spotting those mistakes! I'll put up fixes for those issues soon.

**Pull request merged**